### PR TITLE
Move Scalar Subquery validation logic to the Analyzer

### PR DESCRIPTION
**Pull request by mingmwang**
# Which issue does this PR close?

Closes #.

# Rationale for this change
Leverage the `out reference columns `to do validation, and join filter extraction. Avoid duplicate logic among the decorrelation rules.

# What changes are included in this PR?

1)  Move the Scalar Subquery validation logic from the rule `ScalarSubqueryToJoin` to Analyzer
2) Add more validations for Scalar Subquery
3) Clean up join filter extraction logic in the rule `ScalarSubqueryToJoin`, avoid duplication logic with other rules.

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by mingmwang** [comment]
@jackwener 
Would you please help to take a look at this PR?

**Comment by alamb** [comment]
I am reviewing this now

File: datafusion/optimizer/src/analyzer/subquery.rs [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+        .expressions()
+        .iter()
+        .any(|expr| expr.contains_outer())
+}
+
+fn check_aggregation_in_scalar_subquery(
+    inner_plan: &LogicalPlan,
+    agg: &Aggregate,
+) -> Result<()> {
+    if agg.aggr_expr.is_empty() {
```
Comment (alamb) [inline datafusion/optimizer/src/analyzer/subquery.rs:240]: it seems like you might be able to include this logic in `LogicalPlan::max_rows()` and avoid having a special check for aggregates

File: datafusion/optimizer/src/decorrelate_where_in.rs [hunk 2]
```diff
@@ -557,7 +548,7 @@ mod tests {
         let sq = Arc::new(
             LogicalPlanBuilder::from(scan_tpch_table("orders"))
                 .filter(
-                    col("customer.c_custkey")
+                    out_ref_col(DataType::Int64, "customer.c_custkey")
```
Comment (alamb) [inline datafusion/optimizer/src/decorrelate_where_in.rs:551]: I do like the idea that when creating expressions directly outer references must be annotated explicitly

File: datafusion/optimizer/src/decorrelate_where_in.rs [hunk 3]
```diff
@@ -602,7 +596,7 @@ mod tests {
             .build()?;
 
         let expected = "Projection: customer.c_custkey [c_custkey:Int64]\
-        \n  LeftSemi Join:  Filter: customer.c_custkey = __correlated_sq_1.o_custkey AND customer.c_custkey = customer.c_custkey [c_custkey:Int64, c_name:Utf8]\
```
Comment (alamb) [inline datafusion/optimizer/src/decorrelate_where_in.rs:605]: this seems like an improvement to me

**Comment by yukkit** [comment]
I will help review this PR tomorrow morning

File: datafusion/optimizer/src/analyzer/subquery.rs [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+                        .filter(|max_row| *max_row <= 1)
+                        .is_some()
+                    {
+                        Ok(())
+                    } else {
+                        Err(DataFusionError::Plan(
+                            "Correlated scalar subquery must be aggregated to return at most one row"
+                                .to_string(),
+                        ))
+                    }
```
Comment (yukkit) [inline datafusion/optimizer/src/analyzer/subquery.rs:76]: IMO. When the number of returned results is unknown, can the checking logic be put into execution，I have seen that some databases do this, similar to presto. Of course, this has little effect.

File: datafusion/core/tests/sql/subqueries.rs [hunk 5]
```diff
@@ -203,7 +203,202 @@ async fn subquery_not_allowed() -> Result<()> {
...
+    let err = dataframe.into_optimized_plan().err().unwrap();
+
+    assert_eq!(
+        r#"Context("check_analyzed_plan", Plan("A GROUP BY clause in a scalar correlated subquery cannot contain non-correlated columns"))"#,
+        &format!("{err:?}")
+    );
+
+    Ok(())
+}
+
```
Comment (yukkit) [inline datafusion/core/tests/sql/subqueries.rs:390]: An `aggregated_correlated_scalar_subquery_with_extra_group_by_constant` case could be added

**Review by yukkit — APPROVED** [review]

File: datafusion/expr/src/logical_plan/plan.rs [hunk 6]
```diff
@@ -484,6 +484,97 @@ impl LogicalPlan {
...
+            LogicalPlan::Filter(Filter { input, .. }) => input.max_rows(),
+            LogicalPlan::Window(Window { input, .. }) => input.max_rows(),
+            LogicalPlan::Aggregate(Aggregate {
+                input, group_expr, ..
+            }) => {
+                if group_expr.is_empty() {
+                    Some(1)
+                } else {
+                    input.max_rows()
+                }
```
[the remark this answers is not in the document]
Comment (mingmwang) [reply datafusion/expr/src/logical_plan/plan.rs:501]: Good point. I will try add this in this PR.

File: datafusion/expr/src/logical_plan/plan.rs [hunk 7]
```diff
@@ -484,6 +484,97 @@ impl LogicalPlan {
...
+            | LogicalPlan::CreateCatalog(_)
+            | LogicalPlan::DropTable(_)
+            | LogicalPlan::DropView(_)
+            | LogicalPlan::Explain(_)
+            | LogicalPlan::Analyze(_)
+            | LogicalPlan::Dml(_)
+            | LogicalPlan::DescribeTable(_)
+            | LogicalPlan::Prepare(_)
+            | LogicalPlan::Statement(_)
+            | LogicalPlan::Values(_)
```
[the remark this answers is not in the document]
Comment (mingmwang) [reply datafusion/expr/src/logical_plan/plan.rs:574]: Will add.

File: datafusion/expr/src/logical_plan/plan.rs [hunk 8]
```diff
@@ -484,6 +484,97 @@ impl LogicalPlan {
...
+                ..
+            }) => match join_type {
+                JoinType::Inner | JoinType::Left | JoinType::Right | JoinType::Full => {
+                    match (left.max_rows(), right.max_rows()) {
+                        (Some(left_max), Some(right_max)) => {
+                            let min_rows = match join_type {
+                                JoinType::Left => left_max,
+                                JoinType::Right => right_max,
+                                JoinType::Full => left_max + right_max,
+                                _ => 0,
```
[the remark this answers is not in the document]
Comment (jackwener) [reply datafusion/expr/src/logical_plan/plan.rs:526]: Agree it

[the remark this answers is not in the document]
Comment (mingmwang) [reply datafusion/expr/src/logical_plan/plan.rs:526]: Why it is `None`(unknown) for inner join ? I think it is 0. 
In the case the left max and right max are already known, if the inner join prune all the results(no match tuples), the `min_row` should be 0.

File: datafusion/expr/src/logical_plan/plan.rs [hunk 9]
```diff
@@ -484,6 +484,97 @@ impl LogicalPlan {
...
+                .map(|plan| plan.max_rows())
+                .try_fold(0usize, |mut acc, input_max| {
+                    if let Some(i_max) = input_max {
+                        acc += i_max;
+                        Some(acc)
+                    } else {
+                        None
+                    }
+                }),
+            LogicalPlan::TableScan(TableScan { fetch, .. }) => *fetch,
```
[the remark this answers is not in the document]
Comment (mingmwang) [reply datafusion/expr/src/logical_plan/plan.rs:554]: I think this `fetch` is set due to `limit` push down, if the `limit` is accurate,  I think it is OK
for the scan nodes to return more than required.

File: datafusion/optimizer/src/analyzer/subquery.rs (hunk 4 above)
Comment (mingmwang) [reply datafusion/optimizer/src/analyzer/subquery.rs:76]: Yes, you are right. 
Some of the systems introduced additional operator like `MaxNumRows` operator to do the runtime check.  
I'm not very familiar with Presto/Trino, would you please share me the code ?

File: datafusion/optimizer/src/analyzer/subquery.rs (hunk 1 above)
Comment (mingmwang) [reply datafusion/optimizer/src/analyzer/subquery.rs:240]: Yes.

File: datafusion/optimizer/src/analyzer/subquery.rs [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+        }
+    } else {
+        false
+    }
+}
+
+/// Check whether the window expressions contain a mixture of out reference columns and inner columns
+fn check_mixed_out_refer_in_window(window: &Window) -> Result<()> {
+    let mixed = window.window_expr.iter().any(|win_expr| {
+        win_expr.contains_outer() && !win_expr.to_columns().unwrap().is_empty()
```
[the remark this answers is not in the document]
Comment (mingmwang) [reply datafusion/optimizer/src/analyzer/subquery.rs:330]: I'm afraid I can not use the `?` inside the closure `iter().any(|| )`

File: datafusion/core/tests/sql/subqueries.rs (hunk 5 above)
Comment (mingmwang) [reply datafusion/core/tests/sql/subqueries.rs:390]: Sure, will add.

File: datafusion/optimizer/src/test/mod.rs [hunk 11]
```diff
@@ -171,6 +190,23 @@ pub fn assert_optimized_plan_eq_display_indent(
...
+) {
+    let optimizer = Optimizer::with_rules(rules);
+    let mut optimized_plan = plan.clone();
+    for rule in &optimizer.rules {
+        optimized_plan = optimizer
+            .optimize_recursively(rule, &optimized_plan, &OptimizerContext::new())
+            .expect("failed to optimize plan")
+            .unwrap_or_else(|| optimized_plan.clone());
+    }
+    let formatted_plan = format!("{}", optimized_plan.display_indent_schema());
```
[the remark this answers is not in the document]
Comment (mingmwang) [reply datafusion/optimizer/src/test/mod.rs]: Done

File: datafusion/expr/src/logical_plan/plan.rs [hunk 12]
```diff
@@ -484,6 +484,97 @@ impl LogicalPlan {
             _ => Ok(self),
         }
     }
+
+    /// Returns the maximum number of rows that this plan can output.
```
[the remark this answers is not in the document]
Comment (mingmwang) [reply datafusion/expr/src/logical_plan/plan.rs]: Done

File: datafusion/expr/src/logical_plan/plan.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (mingmwang) [reply datafusion/expr/src/logical_plan/plan.rs]: Done

File: datafusion/optimizer/src/test/mod.rs [hunk 13]
```diff
@@ -134,6 +134,25 @@ pub fn assert_analyzed_plan_eq_display_indent(
...
+    let analyzed_plan =
+        Analyzer::with_rules(rules).execute_and_check(plan, &options, |_, _| {});
+    match analyzed_plan {
+        Ok(plan) => assert_eq!(format!("{}", plan.display_indent()), "An error"),
+        Err(ref e) => {
+            let actual = format!("{e}");
+            if expected.is_empty() || !actual.contains(expected) {
+                assert_eq!(actual, expected)
+            }
+        }
```
[the remark this answers is not in the document]
Comment (mingmwang) [reply datafusion/optimizer/src/test/mod.rs]: Done

File: datafusion/core/tests/sql/subqueries.rs (hunk 5 above)
Comment (mingmwang) [reply datafusion/core/tests/sql/subqueries.rs:390]: Done

File: datafusion/optimizer/src/decorrelate_where_in.rs (hunk 3 above)
Comment (mingmwang) [reply datafusion/optimizer/src/decorrelate_where_in.rs:605]: Yes, a redundant filter was removed.

**Comment by alamb** [comment]
🚀

File: datafusion/optimizer/src/analyzer/subquery.rs (hunk 4 above)
Comment (yukkit) [reply datafusion/optimizer/src/analyzer/subquery.rs:78]: If it is checked at runtime, additional operators need to be introduced. The current pr is much better than before, and this feature can be implemented in future plans. The code of presto is here https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedScalarSubquery.java

**Comment by mingmwang** [comment]
I will merge this PR and continue working on following improvement.

**Pull request merged**